clients/gateway/gateway.go

```diff
@@ -132,19 +144,21 @@ func (c *Client) doPost(ctx context.Context, url string, data any) (*http.Respon
 	return c.client.Do(req)
 }
 
-func (c *Client) tryDecodeErr(resp io.Reader) error {
-	var gatewayError struct {
-		Code    string `json:"code"`
-		Message string `json:"message"`
-	}
+type ErrorCode string
 
-	if err := json.NewDecoder(resp).Decode(&gatewayError); err != nil {
-		c.log.Errorw("failed to decode gateway error", "err", err)
-	}
+type Error struct {
+	Code    ErrorCode `json:"code"`
+	Message string    `json:"message"`
+}
+
+func (e Error) Error() string {
+	return e.Message
+}
 
-	var err error
-	if gatewayError.Message != "" {
-		err = errors.New(gatewayError.Message)
+func (c *Client) tryDecodeErr(resp io.Reader) error {
+	var gatewayError Error
+	if err := json.NewDecoder(resp).Decode(&gatewayError); err != nil {
+		return err
 	}
-	return err
+	return &gatewayError
```

Comment:
Here is a problem: json unmarshaling fails only if:

- it's malformed json
- types are not the same as in a struct

So if there is empty json or json without same fields as in a struct - unmarshaling will be ok. On caller side: 
```
err = c.tryDecodeErr(resp.Body)
// decoding failed, at least pass info about http code
if err == nil {
    err = fmt.Errorf("received non 200 status code: %d", resp.StatusCode)
}
```
this condition will be always false. I decided to log `...Decode()` failure instead of passing it as an error in API (thought that "received non 200 status code..." is source of problem and meaningful then decoding of failure message itself)

Reply:
Nice catch, thank you! Will try to fix and ping you for another review.

Reply:
How about [this](https://github.com/NethermindEth/juno/compare/2bc4dfc6fd6ce68151725dad80d93a4af2748eaf..e339b7df7cb64df89988cb6ecc15b55082e2d1a0)

Reply:
ping @kirugan